File: scraper/scrape.js
#!/usr/bin/env node
/**
 * ILO Abandoned Seafarers scraper (Node.js + Playwright)
 *
 * Usage:
 *   node scrape.js [--start 1] [--end 1705] [--api http://localhost:3001] [--concurrency 4]
 *   node scrape.js --rescan-open [--api http://localhost:3001] [--concurrency 4]
 *
 * Modes:
 *   Default       Range scan from START to END, then auto-extends beyond END until
 *                 EMPTY_STREAK_LIMIT consecutive empty pages — catches newly added records
 *                 without needing to update --end manually.
 *   --rescan-open Re-scrapes every Unresolved and Disputed record already in the DB to
 *                 pick up status changes. Run periodically alongside the default scan.
 *
 * Flags:
 *   --force       Ingest even if the pre-ingest sanity guard fails (use when a
 *                 large drop in record count or field fill-rate is legitimate).
 *
 * Robustness:
 *   Each page is retried up to MAX_RETRIES times with exponential backoff before
 *                 being skipped, preventing transient network errors from causing gaps.
 *   Before ingesting, a sanity guard compares the run to the current snapshot and
 *                 refuses to overwrite good data if the scrape looks broken (0 records,
 *                 record count cratered, or a monitored field's fill-rate collapsed).
 *
 * First run:
 *   npm install
 *   npx playwright install chromium
 */

const { chromium } = require('playwright');
const https = require('https');
const http  = require('http');
const fs    = require('fs');
const path  = require('path');
const { geocode, loadPortOverrides } = require('./geocode');

// ---------------------------------------------------------------------------
// CLI args
// ---------------------------------------------------------------------------
const args = process.argv.slice(2);
function getArg(flag, fallback) {
  const i = args.indexOf(flag);
  return i !== -1 && args[i + 1] ? args[i + 1] : fallback;
}
function hasFlag(flag) { return args.includes(flag); }

const RESCAN_OPEN  = hasFlag('--rescan-open');
const FORCE        = hasFlag('--force'); // bypass the pre-ingest sanity guard
const START        = parseInt(getArg('--start', '1'), 10);
const END          = parseInt(getArg('--end', '1705'), 10);
const API_URL      = getArg('--api', 'http://localhost:3001');
const CONCURRENCY  = parseInt(getArg('--concurrency', '4'), 10);

const MAX_RETRIES        = 3;
const RETRY_BASE_MS      = 3000;
const EMPTY_STREAK_LIMIT = 30; // auto-extend stops after this many consecutive empty pages

// Pre-ingest sanity guard thresholds. A scrape that silently breaks (server
// outage → 0 records; a changed selector → a field blanks out) must not be
// allowed to overwrite good data. These bounds decide when a run looks broken.
const COUNT_FLOOR_RATIO = 0.5; // range scan must return ≥50% of the previous record count
const FILL_MIN_PREV     = 0.5; // only monitor fields populated on >50% of previous records
const FILL_DROP_RATIO   = 0.5; // …and flag them if their fill-rate drops below half its prior value
// Fields expected to be populated on essentially every record regardless of
// status or subset — a collapse here means a broken selector, not real data.
const MONITORED_FIELDS  = ['ship_name', 'port_of_abandonment', 'comments'];

const BASE_URL = 'https://wwwex.ilo.org/dyn/r/abandonment/seafarers/details';

// ---------------------------------------------------------------------------
// Support data loaders  (port geocoding/overrides live in ./geocode)
// ---------------------------------------------------------------------------
function loadFlagUrls() {
  const csv = path.join(__dirname, 'flag_urls.csv');
  if (!fs.existsSync(csv)) return {};
  const lines = fs.readFileSync(csv, 'utf8').split('\n');
  if (lines.length < 2) return {};
  const headers   = lines[0].split(',').map(h => h.trim().toLowerCase());
  const countryIdx = headers.findIndex(h => h === 'country');
  const urlIdx     = headers.findIndex(h => h.includes('url'));
  const flags = {};
  for (const line of lines.slice(1)) {
    const cols    = line.split(',');
    const country = cols[countryIdx]?.trim();
    const url     = cols[urlIdx]?.trim();
    if (country && url) flags[country] = url;
  }
  return flags;
}

// ---------------------------------------------------------------------------
// HTTP helpers
// ---------------------------------------------------------------------------
function fetchJson(url, extraHeaders = {}) {
  return new Promise((resolve, reject) => {
    const mod = url.startsWith('https') ? https : http;
    mod.get(url, { headers: { Accept: 'application/json', ...extraHeaders } }, res => {
      let data = '';
      res.on('data', c => data += c);
      res.on('end', () => { try { resolve(JSON.parse(data)); } catch { reject(new Error('JSON parse')); } });
    }).on('error', reject);
  });
}

function postJson(url, payload) {
  return new Promise((resolve, reject) => {
    const body   = JSON.stringify(payload);
    const parsed = new URL(url);
    const mod    = parsed.protocol === 'https:' ? https : http;
    const req    = mod.request({
      hostname: parsed.hostname,
      port:     parsed.port || (parsed.protocol === 'https:' ? 443 : 80),
      path:     parsed.pathname + parsed.search,
      method:   'POST',
      headers:  { 'Content-Type': 'application/json', 'Content-Length': Buffer.byteLength(body) },
    }, res => {
      let data = '';
      res.on('data', c => data += c);
      res.on('end', () => { try { resolve({ status: res.statusCode, body: JSON.parse(data) }); } catch { resolve({ status: res.statusCode, body: data }); } });
    });
    req.on('error', reject);
    req.write(body);
    req.end();
  });
}

function sleep(ms) { return new Promise(r => setTimeout(r, ms)); }

// ---------------------------------------------------------------------------
// Extract the most recent date from a comments string (format: "DD Month YYYY:")
// Returns ISO date string (YYYY-MM-DD) or null
// ---------------------------------------------------------------------------
const MONTH_MAP = {
  january:0, february:1, march:2, april:3, may:4, june:5,
  july:6, august:7, september:8, october:9, november:10, december:11,
};

function parseLastActivityDate(comments) {
  if (!comments) return null;
  const re = /\b(\d{1,2})\s+(January|February|March|April|May|June|July|August|September|October|November|December)\s+(\d{4})\s*:/gi;
  let latest = null;
  let match;
  while ((match = re.exec(comments)) !== null) {
    const d = new Date(parseInt(match[3], 10), MONTH_MAP[match[2].toLowerCase()], parseInt(match[1], 10));
    if (!latest || d > latest) latest = d;
  }
  if (!latest) return null;
  return `${latest.getFullYear()}-${String(latest.getMonth() + 1).padStart(2, '0')}-${String(latest.getDate()).padStart(2, '0')}`;
}

// ---------------------------------------------------------------------------
// Extract data directly from APEX DOM fields using Playwright evaluate
// ---------------------------------------------------------------------------
function extractApexFields() {
  function val(id) {
    const el = document.getElementById(id);
    return el ? (el.value || '').trim() : null;
  }
  function txt(selector) {
    const el = document.querySelector(selector);
    return el ? (el.innerText || el.textContent || '').trim() : null;
  }
  // The "Comments and observations" content lives in an APEX region whose
  // generated id (e.g. #R...) changes whenever ILO rebuilds the app. Locate it
  // by its visible region title instead of a hardcoded id, then read the
  // dynamic-content body inside it — far more resilient to redeploys.
  function commentsText() {
    const title = [...document.querySelectorAll('.t-Region-title, h2')]
      .find(h => /comments and observations/i.test(h.innerText || ''));
    if (!title) return '';
    const region = title.closest('.t-Region');
    const dyn = region && region.querySelector('a-dynamic-content, .a-dynamic-content');
    return dyn ? (dyn.innerText || dyn.textContent || '').trim() : '';
  }

  const shipName = val('P3_SHIP_NAME');
  if (!shipName) return null;

  const flagRaw = val('P3_FLAG') || '';
  const imoRaw  = val('P3_IMO_NO') || '';
  const imo     = imoRaw.replace(/\D/g, '');

  return {
    ship_name:           shipName,
    ship_status:         val('P3_CASE_STATUS') || '',
    flag:                flagRaw,
    imo_number:          imoRaw,
    vessel_type:         val('P3_IMO_SHIP_TYPE') || '',
    port_of_abandonment: val('P3_PORT') || '',
    abandonment_date:    val('P3_ABAND_DATE') || '',
    notification_date:   val('P3_NOTIF_DATE') || '',
    reporting_member:    val('P3_REPORTING') || '',
    num_seafarers:       parseInt(val('P3_SEAFARERS_NUMBER') || '', 10) || null,
    circumstances:       val('P3_CIRCUMSTANCES') || '',
    comments:            commentsText(),
    vessel_finder_url:   imo ? `https://www.vesselfinder.com/?imo=${imo}` : null,
  };
}

// ---------------------------------------------------------------------------
// Scrape one page — throws on network/timeout/HTTP-5xx/429 error (retriable),
// returns null for a genuinely empty page (200 with no record)
// ---------------------------------------------------------------------------
async function scrapeOne(page, id, portOverrides, flagUrls) {
  const url = `${BASE_URL}?p3_abandonment_id=${id}`;
  const response = await page.goto(url, { waitUntil: 'networkidle', timeout: 25000 });

  // page.goto resolves even on HTTP error statuses, so a Cloudflare 502/503/504
  // (ILO origin down) or 429 (rate limited) would otherwise be parsed as a page
  // with no form fields and silently counted as an empty record. Treat these as
  // retriable errors so scrapeWithRetry backs off and retries instead.
  const status = response ? response.status() : 0;
  if (status === 0 || status === 429 || status >= 500) {
    throw new Error(`HTTP ${status || 'no response'} — server/gateway error`);
  }

  const fields = await page.evaluate(extractApexFields);

  if (!fields) return null;

  const { lat, lon } = await geocode(fields.port_of_abandonment, portOverrides);

  return {
    abandonment_id:      String(id),
    ilo_url:             `${BASE_URL}?p3_abandonment_id=${id}`,
    ship_name:           fields.ship_name,
    ship_status:         fields.ship_status,
    flag:                fields.flag,
    imo_number:          fields.imo_number,
    port_of_abandonment: fields.port_of_abandonment,
    port_latitude:       lat,
    port_longitude:      lon,
    abandonment_date:    fields.abandonment_date,
    notification_date:   fields.notification_date,
    reporting_member:    fields.reporting_member,
    num_seafarers:       fields.num_seafarers,
    circumstances:       fields.circumstances,
    comments:            fields.comments,
    fishing_vessel:      /fishing/i.test(fields.vessel_type) ? 1 : 0,
    vessel_finder_url:   fields.vessel_finder_url,
    flag_url:            flagUrls[fields.flag] || null,
    last_activity_date:  parseLastActivityDate(fields.comments),
  };
}

// ---------------------------------------------------------------------------
// Scrape one ID with retries — always returns null on permanent failure
// ---------------------------------------------------------------------------
async function scrapeWithRetry(browser, id, portOverrides, flagUrls) {
  for (let attempt = 1; attempt <= MAX_RETRIES; attempt++) {
    const page = await browser.newPage();
    try {
      const result = await scrapeOne(page, id, portOverrides, flagUrls);
      await page.close();
      return result; // null = empty page (not an error)
    } catch (e) {
      await page.close();
      if (attempt < MAX_RETRIES) {
        const delay = RETRY_BASE_MS * attempt;
        console.log(`  [${id}] attempt ${attempt}/${MAX_RETRIES} failed — retrying in ${delay / 1000}s (${e.message.split('\n')[0]})`);
        await sleep(delay);
      } else {
        console.log(`  [${id}] FAILED after ${MAX_RETRIES} attempts — skipping`);
      }
    }
  }
  return null;
}

// ---------------------------------------------------------------------------
// Run batches over a list of IDs, logging hits and misses
// ---------------------------------------------------------------------------
async function runBatches(browser, ids, portOverrides, flagUrls) {
  const records = [];
  for (let i = 0; i < ids.length; i += CONCURRENCY) {
    const batch   = ids.slice(i, i + CONCURRENCY);
    const results = await Promise.all(
      batch.map(id => scrapeWithRetry(browser, id, portOverrides, flagUrls))
    );
    for (let j = 0; j < batch.length; j++) {
      const r = results[j];
      if (r) {
        records.push(r);
        console.log(`  [${r.abandonment_id}] ${r.ship_name} — ${r.port_of_abandonment}`);
      } else {
        process.stdout.write('.');
      }
    }
    await sleep(500);
  }
  return records;
}

// ---------------------------------------------------------------------------
// Auto-extend: probe IDs beyond END until EMPTY_STREAK_LIMIT consecutive misses
// ---------------------------------------------------------------------------
async function autoExtend(browser, fromId, portOverrides, flagUrls) {
  console.log(`\nAuto-extending from ID ${fromId} (stops after ${EMPTY_STREAK_LIMIT} consecutive empty pages)…`);
  const records = [];
  let id           = fromId;
  let emptyStreak  = 0;

  while (emptyStreak < EMPTY_STREAK_LIMIT) {
    const batch   = Array.from({ length: CONCURRENCY }, (_, i) => id + i);
    const results = await Promise.all(
      batch.map(bid => scrapeWithRetry(browser, bid, portOverrides, flagUrls))
    );
    for (let j = 0; j < batch.length; j++) {
      const r = results[j];
      if (r) {
        records.push(r);
        emptyStreak = 0;
        console.log(`  [${r.abandonment_id}] ${r.ship_name} — ${r.port_of_abandonment}`);
      } else {
        emptyStreak++;
        process.stdout.write('.');
      }
    }
    id += CONCURRENCY;
    await sleep(500);
  }

  console.log(`\nAuto-extend complete — ${records.length} new record(s) found.`);
  return records;
}

// ---------------------------------------------------------------------------
// Rescan-open mode: fetch Unresolved + Disputed IDs from the API, re-scrape
// ---------------------------------------------------------------------------
async function getOpenIds() {
  const [unresolved, disputed] = await Promise.all([
    fetchJson(`${API_URL}/api/ships?status=Unresolved`),
    fetchJson(`${API_URL}/api/ships?status=Disputed`),
  ]);
  const ids = [...unresolved, ...disputed].map(s => parseInt(s.abandonment_id, 10));
  return [...new Set(ids)].sort((a, b) => a - b);
}

// ---------------------------------------------------------------------------
// Pre-ingest sanity guard — refuse to overwrite good data with a broken scrape
// ---------------------------------------------------------------------------
function fillRate(records, field) {
  if (!records.length) return 0;
  const filled = records.filter(r => {
    const v = r[field];
    return v !== null && v !== undefined && String(v).trim() !== '';
  }).length;
  return filled / records.length;
}

// Fetch the current latest snapshot to compare against. Returns [] if the API
// is unreachable or empty (e.g. a genuine first run) so the guard degrades to
// the 0-records check only rather than blocking on a missing baseline.
async function fetchPreviousSnapshot() {
  try {
    const ships = await fetchJson(`${API_URL}/api/ships`);
    return Array.isArray(ships) ? ships : [];
  } catch {
    return [];
  }
}

// Returns { ok, reasons }. ok=false means the run looks broken and must not be
// ingested (unless --force). prev is the previous snapshot (may be empty).
function sanityCheck(records, prev) {
  const reasons = [];

  // A scrape that produced nothing is always broken — catches total failures
  // (server down, route/param renamed) even with no baseline to compare to.
  if (records.length === 0) {
    return { ok: false, reasons: ['scrape produced 0 records'] };
  }

  // Record-count floor. Skipped for rescan-open, which intentionally scrapes
  // only the open-case subset and would always fall short of the full snapshot.
  if (!RESCAN_OPEN && prev.length) {
    const floor = Math.floor(prev.length * COUNT_FLOOR_RATIO);
    if (records.length < floor) {
      reasons.push(`record count ${records.length} < ${Math.round(COUNT_FLOOR_RATIO * 100)}% of previous snapshot (${prev.length}, floor ${floor})`);
    }
  }

  // Per-field fill-rate collapse — catches a selector that broke and silently
  // blanked a field (the "Comments and observations" region-id change).
  if (prev.length) {
    for (const f of MONITORED_FIELDS) {
      const prevRate = fillRate(prev, f);
      const newRate  = fillRate(records, f);
      if (prevRate >= FILL_MIN_PREV && newRate < prevRate * FILL_DROP_RATIO) {
        reasons.push(`field "${f}" fill-rate collapsed: ${(prevRate * 100).toFixed(0)}% → ${(newRate * 100).toFixed(0)}%`);
      }
    }
  }

  return { ok: reasons.length === 0, reasons };
}

// ---------------------------------------------------------------------------
// Main
// ---------------------------------------------------------------------------
async function main() {
  const portOverrides = loadPortOverrides();
  const flagUrls      = loadFlagUrls();
  const scrapedAt     = new Date().toISOString();

  const browser = await chromium.launch({ headless: true });
  let records   = [];

  if (RESCAN_OPEN) {
    console.log('Mode: rescan-open — re-scraping all Unresolved and Disputed records');
    console.log(`scraped_at: ${scrapedAt}\n`);
    const ids = await getOpenIds();
    console.log(`Fetched ${ids.length} open IDs from API\n`);
    records = await runBatches(browser, ids, portOverrides, flagUrls);
  } else {
    console.log(`Mode: range scan — IDs ${START}–${END} + auto-extend | concurrency=${CONCURRENCY}`);
    console.log(`scraped_at: ${scrapedAt}\n`);
    const ids = Array.from({ length: END - START + 1 }, (_, i) => START + i);
    records = await runBatches(browser, ids, portOverrides, flagUrls);
    const extended = await autoExtend(browser, END + 1, portOverrides, flagUrls);
    records.push(...extended);
  }

  await browser.close();

  console.log(`\nScraped ${records.length} records.`);

  // Guard: don't let a broken scrape overwrite good data. Compare against the
  // current snapshot and bail (saving the raw scrape) if the run looks broken.
  const prev  = await fetchPreviousSnapshot();
  const check = sanityCheck(records, prev);
  if (!check.ok) {
    const out = path.join(__dirname, `scraped_${scrapedAt.slice(0, 10)}.json`);
    fs.writeFileSync(out, JSON.stringify({ scraped_at: scrapedAt, ships: records }, null, 2));
    if (!FORCE) {
      console.error('\n❌ SANITY CHECK FAILED — not ingesting. The scrape looks broken:');
      for (const r of check.reasons) console.error(`   • ${r}`);
      console.error(`\nRaw scrape saved to ${out}`);
      console.error('If this change is legitimate, re-run with --force to ingest anyway.');
      process.exit(1);
    }
    console.warn('\n⚠️  Sanity check failed but --force given — ingesting anyway:');
    for (const r of check.reasons) console.warn(`   • ${r}`);
  }

  console.log(`Posting to ${API_URL}/api/scrapes/ingest …`);
  try {
    const res = await postJson(`${API_URL}/api/scrapes/ingest`, { scraped_at: scrapedAt, ships: records });
    console.log(`Ingest response (${res.status}):`, res.body);
  } catch (e) {
    const out = path.join(__dirname, `scraped_${scrapedAt.slice(0, 10)}.json`);
    fs.writeFileSync(out, JSON.stringify({ scraped_at: scrapedAt, ships: records }, null, 2));
    console.error(`API error: ${e.message}\nFallback saved to ${out}`);
  }
}

main().catch(err => { console.error(err); process.exit(1); });

File: scraper/scrape.py
#!/usr/bin/env python3
"""
ILO Abandoned Seafarers scraper.

Uses Playwright to render each detail page (AJAX site) and extract structured data.
Iterates abandonment IDs 1 through MAX_ID, skipping missing records gracefully.
Geocodes ports via Nominatim (with manual override CSV for corrections).
Posts the full batch to the backend /api/scrapes/ingest endpoint.

Usage:
    pip install -r requirements.txt
    playwright install chromium
    python scrape.py [--start 1] [--end 1700] [--api http://localhost:3001]
"""

import argparse
import csv
import json
import os
import re
import time
from datetime import datetime, timezone
from pathlib import Path

import requests
from geopy.geocoders import Nominatim
from geopy.exc import GeocoderTimedOut, GeocoderUnavailable
from playwright.sync_api import sync_playwright, TimeoutError as PlaywrightTimeout

# ---------------------------------------------------------------------------
# Config
# ---------------------------------------------------------------------------
BASE_URL = "https://wwwex.ilo.org/dyn/r/abandonment/seafarers/details"
SCRIPT_DIR = Path(__file__).parent
REPO_ROOT = SCRIPT_DIR.parent

PORT_OVERRIDES_CSV = REPO_ROOT / "cleaned_ports_list.csv"
FLAG_URLS_CSV = REPO_ROOT / "flag_urls.csv"

GEOCODE_DELAY = 1.2   # seconds between Nominatim calls
PAGE_DELAY = 0.8      # seconds between Playwright page loads


# ---------------------------------------------------------------------------
# Load support data
# ---------------------------------------------------------------------------
def load_port_overrides():
    overrides = {}
    if PORT_OVERRIDES_CSV.exists():
        with open(PORT_OVERRIDES_CSV, newline='', encoding='utf-8') as f:
            reader = csv.DictReader(f, delimiter='~')
            for row in reader:
                overrides[row['Port of abandonment'].strip()] = (
                    float(row['lat']), float(row['long'])
                )
    return overrides


def load_flag_urls():
    flags = {}
    if FLAG_URLS_CSV.exists():
        with open(FLAG_URLS_CSV, newline='', encoding='utf-8') as f:
            reader = csv.DictReader(f)
            for row in reader:
                country = row.get('Country') or row.get('country', '').strip()
                url = row.get('Flag URL') or row.get('flag_url', '').strip()
                if country and url:
                    flags[country] = url
    return flags


# ---------------------------------------------------------------------------
# Geocoding
# ---------------------------------------------------------------------------
_geocoder = Nominatim(user_agent="seafarers_scraper/2.0")
_geocache: dict[str, tuple[float, float] | None] = {}


def geocode(port: str, overrides: dict) -> tuple[float | None, float | None]:
    if not port:
        return None, None
    if port in overrides:
        return overrides[port]
    if port in _geocache:
        result = _geocache[port]
        return result if result else (None, None)

    time.sleep(GEOCODE_DELAY)
    try:
        loc = _geocoder.geocode(port, timeout=10)
        if loc:
            _geocache[port] = (loc.latitude, loc.longitude)
            return loc.latitude, loc.longitude
        _geocache[port] = None
        return None, None
    except (GeocoderTimedOut, GeocoderUnavailable):
        return None, None


# ---------------------------------------------------------------------------
# Page parsing
# ---------------------------------------------------------------------------
def parse_detail_page(html: str, abandonment_id: int) -> dict | None:
    """
    Parse a rendered detail page and return a ship record dict.
    Returns None if the page indicates no record exists.
    """
    from html.parser import HTMLParser

    # Quick check for empty/404 page
    if 'No data found' in html or 'p3_abandonment_id' not in html:
        return None

    def text_after_label(label: str) -> str:
        """Extract the value following a label in the DL/DD structure."""
        pattern = re.compile(
            rf'{re.escape(label)}\s*</(?:dt|th|label)[^>]*>\s*<(?:dd|td)[^>]*>\s*(.*?)\s*</(?:dd|td)>',
            re.IGNORECASE | re.DOTALL,
        )
        m = pattern.search(html)
        if m:
            # Strip inner HTML tags
            return re.sub(r'<[^>]+>', '', m.group(1)).strip()
        return ''

    def find_field(label: str) -> str:
        val = text_after_label(label)
        if not val:
            # Try simpler adjacent text pattern
            idx = html.lower().find(label.lower())
            if idx != -1:
                snippet = html[idx:idx+500]
                tags_stripped = re.sub(r'<[^>]+>', ' ', snippet)
                parts = [p.strip() for p in tags_stripped.split() if p.strip()]
                # Skip the label words themselves
                label_words = label.split()
                for i, p in enumerate(parts):
                    if p not in label_words and len(p) > 1:
                        return p
        return val

    # Extract structured fields via common label patterns
    record = {
        'abandonment_id': str(abandonment_id),
        'ilo_url': f"https://wwwex.ilo.org/dyn/r/abandonment/seafarers/details?p3_abandonment_id={abandonment_id}",
    }

    # Try to pull a table of key-value pairs
    rows = re.findall(
        r'<tr[^>]*>.*?<(?:td|th)[^>]*>(.*?)</(?:td|th)>.*?<(?:td|th)[^>]*>(.*?)</(?:td|th)>.*?</tr>',
        html, re.DOTALL | re.IGNORECASE,
    )

    kv = {}
    for k_raw, v_raw in rows:
        k = re.sub(r'<[^>]+>', '', k_raw).strip().rstrip(':')
        v = re.sub(r'<[^>]+>', ' ', v_raw).strip()
        v = re.sub(r'\s+', ' ', v)
        if k:
            kv[k] = v

    def get(*keys):
        for k in keys:
            for stored_key, val in kv.items():
                if k.lower() in stored_key.lower():
                    return val
        return ''

    record['ship_name']           = get('Ship name', 'Vessel name', 'Name of ship')
    record['ship_status']         = get('Status', 'Ship status')
    record['flag']                = get('Flag', 'Flag state')
    record['imo_number']          = get('IMO', 'IMO number', 'IMO no')
    record['port_of_abandonment'] = get('Port of abandonment', 'Port')
    record['abandonment_date']    = get('Abandonment date', 'Date of abandonment')
    record['notification_date']   = get('Notification date', 'Date of notification')
    record['reporting_member']    = get('Reporting', 'Member', 'Organisation')
    record['num_seafarers']       = get('Number of seafarers', 'No. of seafarers', 'Seafarers')
    record['circumstances']       = get('Circumstances')
    record['comments']            = get('Comments', 'Observations')
    record['fishing_vessel']      = 1 if 'fishing' in get('Fishing', 'Type').lower() else 0

    # Clean num_seafarers to integer
    ns = re.sub(r'[^\d]', '', record['num_seafarers'])
    record['num_seafarers'] = int(ns) if ns else None

    # Build VesselFinder link from IMO
    imo = re.sub(r'[^\d]', '', record['imo_number'])
    record['vessel_finder_url'] = f"https://www.vesselfinder.com/vessels?name={imo}" if imo else None

    return record


# ---------------------------------------------------------------------------
# Main scraper
# ---------------------------------------------------------------------------
def scrape(start: int, end: int, api_url: str):
    port_overrides = load_port_overrides()
    flag_urls = load_flag_urls()
    scraped_at = datetime.now(timezone.utc).isoformat()

    print(f"Starting scrape: IDs {start}–{end}  |  scraped_at={scraped_at}")

    records = []
    failed_ids = []

    with sync_playwright() as pw:
        browser = pw.chromium.launch(headless=True)
        context = browser.new_context(
            user_agent=(
                "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) "
                "AppleWebKit/537.36 (KHTML, like Gecko) "
                "Chrome/124.0.0.0 Safari/537.36"
            )
        )
        page = context.new_page()

        for aid in range(start, end + 1):
            url = f"{BASE_URL}?p3_abandonment_id={aid}"
            try:
                page.goto(url, wait_until='networkidle', timeout=20_000)
                html = page.content()
            except PlaywrightTimeout:
                print(f"  [{aid}] TIMEOUT — skipping")
                failed_ids.append(aid)
                continue
            except Exception as e:
                print(f"  [{aid}] ERROR: {e} — skipping")
                failed_ids.append(aid)
                continue

            record = parse_detail_page(html, aid)
            if record is None:
                print(f"  [{aid}] no record")
                time.sleep(PAGE_DELAY)
                continue

            # Geocode
            lat, lon = geocode(record.get('port_of_abandonment', ''), port_overrides)
            record['port_latitude'] = lat
            record['port_longitude'] = lon

            # Flag URL
            record['flag_url'] = flag_urls.get(record.get('flag', ''), None)

            records.append(record)
            print(f"  [{aid}] {record.get('ship_name', '?')} — {record.get('port_of_abandonment', '?')}")
            time.sleep(PAGE_DELAY)

        browser.close()

    print(f"\nScraped {len(records)} records. Posting to {api_url}/api/scrapes/ingest …")

    payload = {'scraped_at': scraped_at, 'ships': records}
    try:
        resp = requests.post(f"{api_url}/api/scrapes/ingest", json=payload, timeout=60)
        resp.raise_for_status()
        print(f"Ingest response: {resp.json()}")
    except Exception as e:
        # Save to disk as fallback
        out = SCRIPT_DIR / f"scraped_{scraped_at[:10]}.json"
        with open(out, 'w') as f:
            json.dump(payload, f, indent=2)
        print(f"API error: {e}\nData saved to {out}")

    if failed_ids:
        print(f"\nFailed IDs (network errors): {failed_ids}")


# ---------------------------------------------------------------------------
if __name__ == '__main__':
    parser = argparse.ArgumentParser(description='Scrape ILO seafarers abandonment database')
    parser.add_argument('--start', type=int, default=1)
    parser.add_argument('--end',   type=int, default=1700)
    parser.add_argument('--api',   default='http://localhost:3001')
    args = parser.parse_args()
    scrape(args.start, args.end, args.api)
